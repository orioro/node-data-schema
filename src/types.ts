import {
  Expression,
  InterpreterList,
  TypeMap,
  TypeAlternatives,
} from '@orioro/expression'
import { ValidationErrorSpec } from '@orioro/validate'
import {
  Node,
  NodeCollector,
  NodeCollectorContext,
} from '@orioro/tree-collect-nodes'
import { ResolverCandidate } from '@orioro/nested-map'

/**
 * @typedef {Expression | Object} UnresolvedSchema
 */
export type UnresolvedSchema =
  | Expression
  | {
      type: string | Expression
      enum?: any[] | Expression
      errors?: { [key: string]: ValidationErrorSpec } | Expression
      [key: string]: any
    }

/**
 * @typedef {Object} ResolvedSchema
 */
export type ResolvedSchema = {
  type: string
  enum?: any[]
  errors?: { [key: string]: ValidationErrorSpec }
  [key: string]: any
}

/**
 * @typedef {Object} ResolveSchemaContext
 */
export type ResolveSchemaContext = {
  resolvers: ResolverCandidate[]
}

export type SchemaResolverExperssionOptions = {
  interpreters?: InterpreterList
  skipKeys?: string[]
  skipKeysNested?: string[]
}

export type ResolveValueContext = {
  resolvers: ResolverCandidate[]
}

export type SchemaEnvOptions = {
  types?: TypeMap | TypeAlternatives
  interpreters?: InterpreterList
  schemaResolvers?: ResolverCandidate[]
  valueResolvers?: ResolverCandidate[]
  validationCollectors?: NodeCollector[]
}

/**
 * @typedef {Object} ValidationSpec
 */
export type ValidationSpec = Node & {
  validationExpression: Expression
}

export type CollectValidationsPathOptions = {
  include?: string[]
  includePattern?: string[]
  skip?: string[]
  skipPattern?: string[]
}

/**
 * @typedef {Object} CollectValidationsContext
 */
export type CollectValidationsContext = NodeCollectorContext & {
  collectors: NodeCollector[]
  pathOptions?: CollectValidationsPathOptions
  resolveSchema: (schema: UnresolvedSchema, value: any) => ResolvedSchema
}

export type ValidateContext = CollectValidationsContext & {
  interpreters: InterpreterList
}

export type ValidateAsyncOptions = CollectValidationsPathOptions & {
  mode: 'serial' | 'parallel'
}

/**
 * @typedef {String} BuiltInType
 */
export type BuiltInType =
  | 'string'
  | 'number'
  | 'boolean'
  | 'object'
  | 'array'
  | 'date'

export { ValidationErrorSpec, ResolverCandidate, NodeCollector }
