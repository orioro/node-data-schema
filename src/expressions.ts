import {
  EvaluationContext,
  Expression,
  $$VALUE,
  evaluate
} from '@orioro/expression'

const defaultGetType = value => undefined

export type GetTypeInterface = (value: any) => (string | void)

export const schemaTypeExpression = (
  getType:GetTypeInterface = defaultGetType
) => (
  context:EvaluationContext,
  valueExp:Expression = $$VALUE
) => {
  const value = evaluate(context, valueExp)
  let type = getType(value)

  if (typeof type === 'string') {
    return type
  } else {

    type = typeof value

    if (type === 'object') {
      if (Array.isArray(type)) {
        return 'list'
      } else {
        return 'map'
      }
    } else {
      return type
    }
  }
}
