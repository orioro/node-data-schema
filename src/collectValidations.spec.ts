import {
  prepareValidate,
  allowValues,
  prohibitValues,
  parallelCases,
} from '@orioro/validate'

import { schemaTypeExpressions, CORE_SCHEMA_TYPES } from './expressions'
import { ResolvedSchema } from './types'

import { ALL_EXPRESSIONS, interpreterList } from '@orioro/expression'

import {
  resolveSchema as resolveSchema_,
  schemaResolverObject,
  schemaResolverArray,
  schemaResolverExpression,
  schemaResolverDefault,
} from './resolveSchema'

import {
  collectValidations,
  validationCollectorObject,
  validationCollectorArray,
  validationCollectorDefault,
  validationCollectorString,
  validationCollectorNumber,
} from './collectValidations'

import { testCases, fnCallLabel, variableName } from '@orioro/jest-util'

const resolveSchema = resolveSchema_.bind(null, {
  resolvers: [
    schemaResolverExpression(),
    schemaResolverObject(),
    schemaResolverArray(),
    schemaResolverDefault(),
  ],
})

const REQUIRED_ERROR = {
  code: 'REQUIRED_ERROR',
  message: 'This value is required',
}

const TYPE_ERROR = {
  code: 'TYPE_ERROR',
  message: 'Invalid type',
}

const { validateSync } = prepareValidate({
  interpreters: interpreterList({
    ...ALL_EXPRESSIONS,
    ...schemaTypeExpressions(CORE_SCHEMA_TYPES),
  }),
})

describe('collectValidations(schema, context) - required / optional', () => {
  test('required', () => {
    const schema = {
      type: 'string',
      required: true,
      errors: {
        required: REQUIRED_ERROR,
        type: TYPE_ERROR,
      },
    }

    const validations = collectValidations(
      {
        collectors: [
          validationCollectorObject(),
          validationCollectorString(),
          validationCollectorDefault(),
        ],
        resolveSchema,
      },
      schema,
      undefined
    )

    expect(validations).toEqual([
      {
        path: '',
        validationExpression: prohibitValues(
          [null, undefined],
          REQUIRED_ERROR,
          ['$if', ['$isSchemaType', 'string'], null, TYPE_ERROR]
        ),
      },
    ])

    const expectations = [
      [null, [REQUIRED_ERROR]],
      [undefined, [REQUIRED_ERROR]],
      [9, [TYPE_ERROR]],
      ['some text', null],
    ]

    expectations.forEach(([input, expected]) => {
      expect(validateSync(validations[0].validationExpression, input)).toEqual(
        expected
      )
    })
  })

  test('optional', () => {
    const schema = {
      type: 'string',
      errors: {
        type: TYPE_ERROR,
      },
    }

    const validations = collectValidations(
      {
        collectors: [
          validationCollectorObject(),
          validationCollectorString(),
          validationCollectorDefault(),
        ],
        resolveSchema,
      },
      schema,
      undefined
    )

    expect(validations).toMatchObject([
      {
        path: '',
        validationExpression: allowValues(
          [null, undefined],
          ['$if', ['$isSchemaType', 'string'], null, { code: 'TYPE_ERROR' }]
        ),
      },
    ])

    const expectations = [
      [null, null],
      [undefined, null],
      [9, [TYPE_ERROR]],
      ['some text', null],
    ]

    expectations.forEach(([input, expected]) => {
      expect(validateSync(validations[0].validationExpression, input)).toEqual(
        expected
      )
    })
  })
})

test('string validations', () => {
  const MIN_LENGTH_ERROR = {
    code: 'MIN_LENGTH_ERROR',
    message: 'Text must have at least 1 char',
  }

  const MAX_LENGTH_ERROR = {
    code: 'MAX_LENGTH_ERROR',
    message: 'Text must have at most 10 chars',
  }

  const schema = {
    type: 'string',
    required: true,
    minLength: 5,
    maxLength: 10,
    pattern: ['^a.+z$', 'i'],
    errors: {
      required: REQUIRED_ERROR,
      type: TYPE_ERROR,
      minLength: MIN_LENGTH_ERROR,
      maxLength: MAX_LENGTH_ERROR,
    },
  }

  const validations = collectValidations(
    {
      collectors: [validationCollectorString(), validationCollectorDefault()],
      resolveSchema,
    },
    schema,
    'Some text'
  )

  expect(validations).toEqual([
    {
      path: '',
      validationExpression: prohibitValues([null, undefined], REQUIRED_ERROR, [
        '$if',
        ['$isSchemaType', 'string'],
        parallelCases([
          [['$gte', 5, ['$stringLength']], MIN_LENGTH_ERROR],
          [['$lte', 10, ['$stringLength']], MAX_LENGTH_ERROR],
        ]),
        TYPE_ERROR,
      ]),
    },
  ])

  const expectations = [
    [null, [REQUIRED_ERROR]],
    [undefined, [REQUIRED_ERROR]],
    [9, [TYPE_ERROR]],
    ['a', [MIN_LENGTH_ERROR]],
    ['abcdefghijklmnopqrstuv', [MAX_LENGTH_ERROR]],
    ['abcdez', null],
    ['AbcdeZ', null],
  ]

  expectations.forEach(([input, expected]) => {
    expect(validateSync(validations[0].validationExpression, input)).toEqual(
      expected
    )
  })
})

describe('array validations', () => {
  const MIN_LENGTH_ERROR = {
    code: 'MIN_LENGTH_ERROR',
  }

  const MAX_LENGTH_ERROR = {
    code: 'MAX_LENGTH_ERROR',
  }

  const schema = {
    type: 'array',
    items: {
      type: 'string',
      required: true,
      minLength: 5,
      maxLength: 10,
      errors: {
        required: REQUIRED_ERROR,
        type: TYPE_ERROR,
        minLength: MIN_LENGTH_ERROR,
        maxLength: MAX_LENGTH_ERROR,
      },
    },
  }

  test('basic', () => {
    const validations = collectValidations(
      {
        collectors: [
          validationCollectorArray(),
          validationCollectorString(),
          validationCollectorDefault(),
        ],
        resolveSchema,
      },
      schema,
      ['123', '12345', '1234567', '123456789012345']
    )

    expect(validations).toMatchSnapshot()
  })
})

describe('object validations', () => {
  const REQUIRED_STRING = {
    type: 'string',
    required: true,
  }

  const REQUIRED_NUMBER = {
    type: 'number',
    required: true,
  }

  const schemaSimple: ResolvedSchema = {
    type: 'object',
    properties: {
      key1: REQUIRED_STRING,
      key2: REQUIRED_STRING,
      key3: REQUIRED_STRING,
      key4: REQUIRED_STRING,
    },
  }

  const schemaComplex: ResolvedSchema = {
    type: 'object',
    properties: {
      key1: REQUIRED_STRING,
      key2: REQUIRED_NUMBER,
      key3: {
        type: 'object',
        properties: {
          key31: REQUIRED_STRING,
          key32: REQUIRED_NUMBER,
        },
      },
      key4: {
        type: 'object',
        properties: {
          key41: REQUIRED_STRING,
          key42: {
            type: 'object',
            properties: {
              key421: REQUIRED_NUMBER,
              key422: REQUIRED_STRING,
            },
          },
          key43: REQUIRED_NUMBER,
        },
      },
    },
  }

  const context = {
    collectors: [
      validationCollectorObject(),
      validationCollectorString(),
      validationCollectorNumber(),
      validationCollectorDefault(),
    ],
    resolveSchema,
  }

  test('basic', () => {
    const validations = collectValidations(context, schemaComplex, {})

    expect(validations).toMatchSnapshot()

    expect(validations.map((v) => v.path)).toEqual([
      '',
      'key1',
      'key2',
      'key3',
      'key3.key31',
      'key3.key32',
      'key4',
      'key4.key41',
      'key4.key42',
      'key4.key42.key421',
      'key4.key42.key422',
      'key4.key43',
    ])
  })

  describe('pathOptions', () => {
    testCases(
      [
        [{ include: ['key1', 'key4'] }, schemaSimple, ['key1', 'key4']],
        [{ include: ['key1', 'key4'] }, schemaComplex, ['key1', 'key4']],
        [
          { includePattern: ['key1', 'key4'] },
          schemaComplex,
          [
            'key1',
            'key4',
            'key4.key41',
            'key4.key42',
            'key4.key42.key421',
            'key4.key42.key422',
            'key4.key43',
          ],
        ],
        [
          { skipPattern: ['key1', 'key4'] },
          schemaComplex,
          ['', 'key2', 'key3', 'key3.key31', 'key3.key32'],
        ],
      ],
      (pathOptions, schema) => {
        return collectValidations(
          {
            ...context,
            pathOptions,
          },
          schema,
          {}
        ).map((v) => v.path)
      },
      ([pathOptions, schema], result) =>
        fnCallLabel(
          'collectValidations - pathOptions',
          [
            pathOptions,
            variableName(
              schema === schemaSimple ? 'schemaSimple' : 'schemaComplex'
            ),
          ],
          result
        )
    )
  })
})
