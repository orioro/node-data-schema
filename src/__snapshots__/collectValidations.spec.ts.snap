// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array validations basic 1`] = `
Array [
  Object {
    "path": "",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      null,
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "array",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`array\`",
        },
      ],
    ],
  },
  Object {
    "path": "0",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
      ],
    ],
  },
  Object {
    "path": "1",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
      ],
    ],
  },
  Object {
    "path": "2",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
      ],
    ],
  },
  Object {
    "path": "3",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "This value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$gte",
                    5,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MIN_LENGTH_ERROR",
                  },
                ],
                Array [
                  Array [
                    "$lte",
                    10,
                    Array [
                      "$stringLength",
                    ],
                  ],
                  Object {
                    "code": "MAX_LENGTH_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Invalid type",
        },
      ],
    ],
  },
]
`;

exports[`object validations basic 1`] = `
Array [
  Object {
    "path": "",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      null,
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "object",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$arrayIncludesAll",
                    Array [
                      "$objectKeys",
                    ],
                    Array [
                      "key1",
                      "key2",
                      "key3",
                      "key4",
                    ],
                  ],
                  Object {
                    "code": "OBJECT_UNKNOWN_PROPERTIES_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`object\`",
        },
      ],
    ],
  },
  Object {
    "path": "key1",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`string\`",
        },
      ],
    ],
  },
  Object {
    "path": "key2",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "number",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`number\`",
        },
      ],
    ],
  },
  Object {
    "path": "key3",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      null,
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "object",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$arrayIncludesAll",
                    Array [
                      "$objectKeys",
                    ],
                    Array [
                      "key31",
                      "key32",
                    ],
                  ],
                  Object {
                    "code": "OBJECT_UNKNOWN_PROPERTIES_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`object\`",
        },
      ],
    ],
  },
  Object {
    "path": "key3.key31",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`string\`",
        },
      ],
    ],
  },
  Object {
    "path": "key3.key32",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "number",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`number\`",
        },
      ],
    ],
  },
  Object {
    "path": "key4",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      null,
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "object",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$arrayIncludesAll",
                    Array [
                      "$objectKeys",
                    ],
                    Array [
                      "key41",
                      "key42",
                      "key43",
                    ],
                  ],
                  Object {
                    "code": "OBJECT_UNKNOWN_PROPERTIES_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`object\`",
        },
      ],
    ],
  },
  Object {
    "path": "key4.key41",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`string\`",
        },
      ],
    ],
  },
  Object {
    "path": "key4.key42",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      null,
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "object",
        ],
        Array [
          "$pipe",
          Array [
            Array [
              "$arrayMap",
              Array [
                "$if",
                Array [
                  "$evaluate",
                  Array [
                    "$value",
                    "0",
                  ],
                  Array [
                    "$value",
                    "$$PARENT_SCOPE",
                  ],
                ],
                null,
                Array [
                  "$value",
                  "1",
                ],
              ],
              Array [
                Array [
                  Array [
                    "$arrayIncludesAll",
                    Array [
                      "$objectKeys",
                    ],
                    Array [
                      "key421",
                      "key422",
                    ],
                  ],
                  Object {
                    "code": "OBJECT_UNKNOWN_PROPERTIES_ERROR",
                  },
                ],
              ],
            ],
            Array [
              "$arrayFilter",
              Array [
                "$notEq",
                null,
              ],
            ],
          ],
        ],
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`object\`",
        },
      ],
    ],
  },
  Object {
    "path": "key4.key42.key421",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "number",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`number\`",
        },
      ],
    ],
  },
  Object {
    "path": "key4.key42.key422",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "string",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`string\`",
        },
      ],
    ],
  },
  Object {
    "path": "key4.key43",
    "validationExpression": Array [
      "$if",
      Array [
        "$in",
        Array [
          null,
          undefined,
        ],
      ],
      Object {
        "code": "REQUIRED_ERROR",
        "message": "Value is required",
      },
      Array [
        "$if",
        Array [
          "$isSchemaType",
          "number",
        ],
        null,
        Object {
          "code": "TYPE_ERROR",
          "message": "Must be type \`number\`",
        },
      ],
    ],
  },
]
`;
